api/v1alpha1/quota_policy.go

```diff
@@ -0,0 +1,121 @@
+// Copyright Envoy AI Gateway Authors
+// SPDX-License-Identifier: Apache-2.0
+// The full text of the Apache license is available in the LICENSE file at
+// the root of the repo.
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	gwapiv1a2 "sigs.k8s.io/gateway-api/apis/v1alpha2"
+)
+
+// QuotaPolicy specifies token quota configuration for inference services.
+// Providing a list of backends in the AIGatewayRouteRule allows failover to a different service
+// if token quota for a service had been exceeded.
+//
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:object:root=true
+// +kubebuilder:subresource:status
+// +kubebuilder:printcolumn:name="Status",type=string,JSONPath=`.status.conditions[-1:].type`
+// +kubebuilder:metadata:labels="gateway.networking.k8s.io/policy=direct"
+type QuotaPolicy struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	Spec              QuotaPolicySpec `json:"spec,omitempty"`
+	// Status defines the status details of the QuotaPolicy.
+	Status QuotaPolicyStatus `json:"status,omitempty"`
+}
+
+// QuotaPolicySpec specifies rules for computing token based costs of requests.
+type QuotaPolicySpec struct {
+	// TargetRefs are the names of the AIServiceBackend resources this QuotaPolicy is being attached to.
+	// Attaching multiple QuotaPolicies to the same AIServiceBackend is invalid and will result in an error
+	// during the reconciliation of AIServiceBackend.
+	//
+	// +optional
+	// +kubebuilder:validation:MaxItems=16
+	// +kubebuilder:validation:XValidation:rule="self.all(ref, ref.group == 'aigateway.envoyproxy.io' && ref.kind == 'AIServiceBackend')", message="targetRefs must reference AIServiceBackend resources"
+	TargetRefs []gwapiv1a2.LocalPolicyTargetReference `json:"targetRefs,omitempty"`
+
+	// LLMRequestCosts specifies how to capture the cost of the LLM-related request, notably the token usage.
+	// The AI Gateway filter will capture each specified number and store it in the Envoy's dynamic
+	// metadata per HTTP request. The namespaced key is "io.envoy.ai_gateway",
+	//
+	// For example, let's say we have the following LLMRequestCosts configuration:
+	// ```yaml
+	//	llmRequestCosts:
+	//	- metadataKey: llm_input_token
+	//	  type: InputToken
+	//	- metadataKey: llm_output_token
+	//	  type: OutputToken
+	//	- metadataKey: llm_total_token
+	//	  type: TotalToken
+	//	- metadataKey: llm_cached_input_token
+	//	  type: CachedInputToken
+	// ```
+	// The token costs computed using this policy are debited from the quota specificed by the BackendTrafficPolicy
+	// attached to the same AIServiceBackend. In the BackendTrafficPolicy you can have multiple buckets
```

Comment:
BackendTrafficPolicy can be attached to the HTTPRoute or gateway level not AIServiceBackend.